src/optionals/queue.c

```diff
@@ -0,0 +1,203 @@
+#include <string.h>
+
+#include "queue.h"
+#include "optionals.h"
+#include "../vm/vm.h"
+
+typedef struct {
+    void **dq;
+    int cap;
+    int front;
+    int rear;
+    int count;
+} Queue;
+
+#define AS_QUEUE(v) ((Queue*)AS_ABSTRACT(v)->data)
+
+void freeQueue(DictuVM *vm, ObjAbstract *abstract) {
+    Queue *queue = (Queue*)abstract->data;
+    FREE(vm, void**, queue->dq);
+    FREE(vm, Queue, abstract->data);
+}
+
+char *queueToString(ObjAbstract *abstract) {
+    UNUSED(abstract);
+
+    char *queueString = malloc(sizeof(char) * 8);
+    snprintf(queueString, 8, "<Queue>");
+    return queueString;
+}
+
+
+static Value queueIsFull(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "size() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+    bool isFull = queue->cap == queue->count;
+
+    return BOOL_VAL(isFull);
+}
+
+static Value queueIsEmpty(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "empty() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+    bool isFull = queue->count == 0;
+
+    return BOOL_VAL(isFull);
+}
+
+static Value queueCap(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "cap() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    return NUMBER_VAL(queue->cap);
+}
+
+static Value queueLen(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "len() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    return NUMBER_VAL(queue->count);
+}
+
+static Value queuePeek(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "peek() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+    double *d = queue->dq[queue->front];
+    
+    return newResultSuccess(vm, NUMBER_VAL(*d));
+}
+
+static Value queuePush(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "push() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    if (queue->cap != queue->count) {
+        if(queue->rear == queue->cap-1) {
+            queue->rear = -1;            
+        }
+
+        double *data = ALLOCATE(vm, double, 1);
+        *data = AS_NUMBER(args[1]);
```

Comment:
Here we are assuming anything pushed to the queue is a number, but in reality we'd want this to be any Value type

Reply:
I forgot to update that. First go was to get numbers working. The weird thing is that all types work.